libraries/chain/controller.cpp

```diff
@@ -546,6 +558,11 @@ struct controller_impl {
 
    void startup(std::function<bool()> shutdown, const genesis_state& genesis) {
       EOS_ASSERT( db.revision() < 1, database_exception, "This version of controller::startup only works with a fresh state database." );
+      const auto& genesis_chain_id = genesis.compute_chain_id();
+      EOS_ASSERT( genesis_chain_id == chain_id, misc_exception,
```

Comment:
There is a `chain_id_type_exception` which might be a better fit. Here and below.